### PR TITLE
parallel for findNodesThatFit and PrioritizeNodes

### PR DESCRIPTION
**Pull request by magicwang-cn**
1. findNodesThatFit  and  PrioritizeNodes are running parallel;
2. logic in both findNodesThatFit  and PrioritizeNodes are running parallel.

ref [19005](https://github.com/kubernetes/kubernetes/issues/19005)

### CODE REVIEW

**Comment by magicwang-cn** [comment]
I signed it!

**Comment by mqliang** [comment]
FYI https://github.com/kubernetes/kubernetes/pull/17400

**Comment by HardySimpson** [comment]
@xiang90 @kevin-wangzefeng @wojtek-t @huangyuqi

**Comment by magicwang-cn** [comment]
@mqliang there are not the same

**Comment by magicwang-cn** [comment]
@davidopp 
can you hava a look at this CI test?

**Comment by xiang90** [comment]
@magicwang-cn I do not feel this is the best approach. We should not start a go routine per node calculation. We do not usually have 1000s of CPUs. Each routine will just be waiting for scheduling by go runtime. This is a waste of resource. What I would suggest is to just partition the nodes to N cpus and run them in parallel.

**Comment by magicwang-cn** [comment]
@xiang90 
sorry, maybe i say not clearly.
not start a goroutine per node, but start a goroutine per priorityConfig(also extenders), you can see the channel array size:
```
func NewScheduleCache(nodeLen, priLen, extenderLen int) *ScheduleCache {
	sCache := &ScheduleCache{
		Size: priLen,
		FResult: FitResult{
			Err:              make(chan error),
			FailedPredicates: FailedPredicateMap{},
		},
		PResult: PriorityResult{
			Err:             make(chan error),
			PrioritizeNodes: schedulerapi.HostPriorityList{},
		},
	}

	sCache.FResult.FailedPredicates = FailedPredicateMap{}
	sCache.PResult.PrioritizeNodes = schedulerapi.HostPriorityList{}

	if extenderLen != 0 {
		sCache.Size += 1
	}

	if sCache.Size == 0 {
		sCache.Size = 1
	}

	for i := 0; i < sCache.Size; i++ {
		sCache.Sub = append(sCache.Sub, make(chan *api.Node, nodeLen))
	}

	return sCache
}
```

File: plugin/pkg/scheduler/generic_scheduler.go [hunk 1]
```diff
@@ -34,15 +33,73 @@ import (
...
+	Err              chan error
 	FailedPredicates FailedPredicateMap
 }
 
+type PriorityResult struct {
+	Err             chan error
+	PrioritizeNodes schedulerapi.HostPriorityList
+}
+
+type ScheduleCache struct {
```
Comment (xiang90) [inline plugin/pkg/scheduler/generic_scheduler.go]: I feel we do not need a cache. We just need a one direction channel that link a output to an input. Let's remove this and use a plain chan instead?

**Comment by hongchaodeng** [comment]
>  Gain about 20% performance up
I suspect the claim.
The data you showed looks like normal deviation. Results on larger scenarios are expected.
At the mean time, I would like to wait optimization issue #18831 to finish first. Optimization like this are pre-mature and more effects w.r.t. mem, cpu interferece should be considered.

**Comment by magicwang-cn** [comment]
@xiang90 @hongchaodeng 
i think this idea is not confilict with #18831(https://github.com/kubernetes/kubernetes/issues/18831) .
i will give the Results on larger scenarios.

**Comment by hongchaodeng** [comment]
Large scenarios benchmark is one point.
More tradeoff weighing on the parallel factors and cache like @xiang90 mentioned above is another. I'm not sure what kind of goal to be achieved here. The tradeoff should be made clear though.

**Comment by magicwang-cn** [comment]
@hongchaodeng @xiang90 @HardySimpson 
before optimized, findNodesThatFit  and PrioritizeNodes just running serial, and with the sequence as follows:
- first, run predicate one by one, and the results(`filtered nodes`) is the second's premise
- second, waiting the first step complete, then use the (`filtered nodes`) to run extender schedulers
- third, waiting for findNodesThatFit  is completed, (`the final filtered nodes`)  as PrioritizeNodes 's premise, then run the priorityFunc parallel.

so as described above, only PrioritizeNodes runs parallel, findNodesThatFit  runs serial and PrioritizeNodes will waiting for findNodesThatFit  finshed.
benchmark result:
```
BenchmarkScheduling100Nodes0Pods	     300	   5939650 ns/op
BenchmarkScheduling100Nodes1000Pods	     200	   7633965 ns/op
BenchmarkScheduling1000Nodes0Pods	     100	  15921910 ns/op
BenchmarkScheduling1000Nodes1000Pods	     100	  18119219 ns/op
BenchmarkScheduling1000Nodes10000Pods	      30	  43473269 ns/op
BenchmarkScheduling5000Nodes10000Pods	      20	 115848593 ns/op
```

i made the following optimization:
- run findNodesThatFit  and PrioritizeNodes parallel, PrioritizeNodes do not need to wait findNodesThatFit  finshed, just use channel to contact each other
- adjust the running sequence of predicate and extender schedulers in findNodesThatFit  , first call extender schedulers(`yes, this may increate nodes data compared to before`), then runs predicate parallel, and each goroutine send the (`filtered node`) to channel;
- change (`type PriorityFunction func(pod *api.Pod, machineToPods map[string][]*api.Pod, podLister PodLister, nodeLister NodeLister) (schedulerapi.HostPriorityList, error)
`) to (`type PriorityFunction func(pod *api.Pod, machineToPods map[string][]*api.Pod, podLister PodLister, nodes chan *api.Node) (schedulerapi.HostPriorityList, error)`), each priorityFunc just read its own nodes data channel.
benchmark result:
```
BenchmarkScheduling100Nodes0Pods	     300	   5811163 ns/op
BenchmarkScheduling100Nodes1000Pods	     200	   7078510 ns/op
BenchmarkScheduling1000Nodes0Pods	     100	  13717007 ns/op
BenchmarkScheduling1000Nodes1000Pods	     100	  15299814 ns/op
BenchmarkScheduling1000Nodes10000Pods	      50	  28764322 ns/op
BenchmarkScheduling5000Nodes10000Pods	      20	  76519593 ns/op
```

**Comment by hongchaodeng** [comment]
The code is lengthy and messy. I judge based on your comments:

> run findNodesThatFit and PrioritizeNodes parallel

Can you describe more about "parallel"

> PrioritizeNodes do not need to wait findNodesThatFit finished, just use channel to contact each other

Why do we need to change that? How is it different from current code? What benefit gains?

> adjust the running sequence of predicate and extender schedulers in findNodesThatFit

Same as above.

> each priorityFunc just read its own nodes data channel

This is duplicate to the second question. Right?

**Comment by HardySimpson** [comment]
Actually this is not a cache, as @xiang90 said. it is just a channel between **predict** and **priority**.
1. when **predict func** get a fit node, put to channel. 
2. At this time **priority func** don't need to wait all fit nodes ready, just caculate the fit node now. 
3. At last all fit node is filtered out by **predict func**, and **priority func** has already caculate 80% of fit nodes
4. **priority func** caculate 20% left node, and sort them, and return the result

**Comment by hongchaodeng** [comment]
All right. I'm very confused. 

> The benefit gains comes from priority func can start before all fit node flitered out

How is it related to parallelization?

> priority func caculate 20% left node, and sort them, and return the result

This sounds pre-mature optimization. How do you know the interface change you made is generic and applied to incoming cases?

Can you guys share what your goal is? What problem you have observed? Describe your proposed change and what benefits it brings, what tradeoff it made?

**Comment by HardySimpson** [comment]
>This sounds pre-mature optimization. How do you know the interface change you made is generic and applied to incoming cases?

20% and 80% is just a metaphor. 
Yes the interface of the priority is changed. But in what case the change do not work?
It works better now. 


>Can you guys share what your goal is? What problem you have observed? Describe your proposed change and what benefits it brings, what tradeoff it made?

The Goal is simple, make scheduler work faster as you did. The benefits you can see in test result. 
Please give us some concrete reason and case to against this change.

**Comment by hongchaodeng** [comment]
The PR is trying to solve many things without description. I'm trying to figure out the goal and explicit reason for code changes. That's not "being against this change".

1. You are trying to change the fits to pass results to priorities via async channel. If so, the parallelization is orthogonal. 

2. Interface change is a big thing. If your change just makes existing code more efficient, that's all green pass. The change you made has put tradeoff on generic interface, cores utilization. 

3. The PR here din't have any issues raised and discussion involved before. It doesn't provide any test/benchmarks for the key change. See [Contributing Guide](https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md#contributing-a-patch)

It's up to @davidopp who handles and accept/rejects PRs. I'm helping review scheduler related code because I need to rebase with huge amount of code. At the mean time, getting the goal clear and commits good is great benefits to entire community.

**Comment by magicwang-cn** [comment]
@hongchaodeng 
sorry, this PR is submitted Irregularities :(, i will make it more formal later,  next time before submitted PR i will do the necessary work [Contributing Guide](https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md#contributing-a-patch).

thank you for your review :).

**Comment by HardySimpson** [comment]
@hongchaodeng thanks your review. Let's back to the code.

https://github.com/kubernetes/kubernetes/pull/19004#issuecomment-166813187

This is the test/benchmarks using your benchmark tools.

**Comment by hongchaodeng** [comment]
@HardySimpson 
I mean the test/benchmark for key path.
For example, if the fits and predicates are supposed to be faster with async channels or parallelization, test/benchmark that only.

**Comment by hongchaodeng** [comment]
@magicwang-cn 
Great. No worries. Feel free to ping me in raising the issue and more code changes.

**Comment by magicwang-cn** [comment]
@davidopp @HardySimpson @kevin-wangzefeng

**Comment by magicwang-cn** [comment]
@davidopp can you review this PR?

**Comment by davidopp** [comment]
I agree it would probably be good to wait on #18831, at least #19090

**Comment by magicwang-cn** [comment]
i will reopen this PR after #18831

**Comment by jayunit100** [comment]
I guess the key thing here is that priority is not a relative value, so thats why we  dont need all nodes to calculate it

**Pull request closed without merging**